src/Fixer/Basic/NumericLiteralSeparatorFixer.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Fixer\Basic;
+
+use PhpCsFixer\AbstractFixer;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\FixerDefinition\FixerDefinitionInterface;
+use PhpCsFixer\Preg;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * Let's you add underscores to numeric literals.
+ *
+ * Inspired by:
+ * - {@link https://github.com/kubawerlos/php-cs-fixer-custom-fixers/blob/main/src/Fixer/NumericLiteralSeparatorFixer.php}
+ * - {@link https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/rules/numeric-separators-style.js}
+ *
+ * @author Marvin Heilemann <11534760+muuvmuuv@users.noreply.github.com>
+ */
+final class NumericLiteralSeparatorFixer extends AbstractFixer
+{
+    private string $separator = '_';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'Adds separators to numeric literals of any kind.',
+            [
+                new CodeSample("<?php \$var = 123456;\n"),
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens): bool
+    {
+        if (\PHP_VERSION_ID < 70400) {
+            // Syntax since PHP 7.4.0
+            return false;
+        }
+
+        return $tokens->isAnyTokenKindsFound([T_DNUMBER, T_LNUMBER]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
+    {
+        foreach ($tokens as $index => $token) {
+            if (!$token->isGivenKind([T_DNUMBER, T_LNUMBER])) {
+                continue;
+            }
+
+            $content = $token->getContent();
+
+            // Already has literal separator, but may be incomplete
+            // e.g.: 1_412412.
+            $content = str_replace('_', '', $content);
+
+            $newContent = $this->formatValue($content);
+
+            if ($content === $newContent) {
+                // Skip Token override if its the same content, like when it
+                // already got a valid literal separator structure.
+                continue;
+            }
+
+            $token = new Token([$token->getId(), $newContent]);
+
+            $tokens[$index] = $token;
+        }
+    }
+
+    private function formatValue(string $value): string
+    {
+        if (str_starts_with(strtolower($value), '0b')) {
+            // Binary
+            return $this->insertEveryRight($value, 4, 2);
```

Comment:
Why add a separator in the middle of a byte?

Reply:
What do you mean? For me this is more readable `0x1001_0010`.

Reply:
Why did you give an example of hexadecimal? here are binary ones split and for them, 8 characters make a byte.

Also when you claim "for me this is more readable" keep in mind it doesn't have to be the same for the rest of humankind :wink: Have you ever seen any tool that groups hexadecimal numbers in a group of 4 characters? Everywhere I look it is either 2 or 8, e.g. [HxD](https://mh-nexus.de/en/hxd/) (see the screenshot there).

Reply:
I just mistyped x → b… and yes others might have other opinions on what is more readable, but I want to avoid starting a "tab vs. spaces war" here. We need a start and then let the community opt-in and provide feedback.

Reply:
So let's start with add adding `_` every byte and then add it in half-byte if anyone will need it.

Reply:
This still is not resolved, it is not "tab vs. spaces war", but "reasonable vs silly" solution - who would ever want to format numbers in byte notation with `_` every half of byte?

Reply:
Ok, I misunderstood you. You were referring to bytes as 8-bit blocks for readability, makes sense. I have mostly worked with 4 (embedded systems) and not seen a lot of doing 8-bit blocks.